Type Improvements for neuron_deep_lift.py and test_neuron_deeplift_basic.py

ashkan-software opened the pull request:
This PR will improve the type improvements for the following 2 files:

- captum/attr/_core/neuron/neuron_deep_lift.py
- tests/attr/neuron/test_neuron_deeplift_basic.py

ashkan-software commented:
Sorry for the large number of commits. I am new to this project :)

All the tests are passing for the files in this PR (`neuron_deep_lift.py` and `test_neuron_deeplift_basic.py`), but fail for some other files:

```
[yousefpour@devvm2221.frc2 ~/captum] flake8 .
captum/attr/_core/deep_lift.py:683:27: C407 Unnecessary list comprehension - 'tuple' can take a generator.
captum/attr/_core/deep_lift.py:689:30: C407 Unnecessary list comprehension - 'tuple' can take a generator.
captum/insights/widget/widget.py:36:27: C408 Unnecessary dict call - rewrite as a literal.
captum/insights/widget/widget.py:46:34: C408 Unnecessary dict call - rewrite as a literal.
tests/attr/test_data_parallel.py:777:23: C416 Unnecessary list comprehension - rewrite using list().
[yousefpour@devvm2221.frc2 ~/captum] ./scripts/run_mypy.sh 
captum/attr/_utils/common.py:95: error: Overloaded function signatures 1 and 2 overlap with incompatible return types
captum/attr/_core/layer/layer_deep_lift.py:293: error: Definition of "attribute" in base class "LayerDeepLift" is incompatible with definition in base class "DeepLiftShap"
[yousefpour@devvm2221.frc2 ~/captum]
```

So I ignored the errors above.

vivekmig commented on tests/attr/neuron/test_neuron_deeplift_basic.py:
```diff
@@ -114,8 +118,22 @@ def test_relu_deepliftshap_with_custom_attr_func(self):
+        expected,
+    ) -> None:
+        def custom_attr_func(
+            multipliers: TensorOrTupleOfTensors,
```
multipliers, inputs, and the return type here should be Tuples of Tensors, while baselines should be a Tuple of either Tensor, int or float

vivekmig replied:
Sorry, I think this might not have been clear, these should just be Tuple[Tensor, ...] not TensorOrTupleOfTensors and baselines should be just Optional[Tuple[Union[Tensor, int, float], ...]]

ashkan-software replied:
Thanks!

vivekmig commented on tests/attr/neuron/test_neuron_deeplift_basic.py:
```diff
@@ -114,8 +118,22 @@ def test_relu_deepliftshap_with_custom_attr_func(self):
-        def custom_attr_func(multipliers, inputs, baselines):
+    def _relu_custom_attr_func_assert(
+        self,
+        attr_method,
```
Can you add type hints to the remaining arguments here?

ashkan-software replied:
Thanks!

ashkan-software commented on captum/attr/_core/neuron/neuron_deep_lift.py:
```diff
@@ -24,13 +29,15 @@ def __init__(self, model, layer):
+        baselines: Optional[
+            Union[Tensor, int, float, Tuple[Union[Tensor, int, float], ...]]
+        ] = None,
+        additional_forward_args: Optional[Any] = None,
```
Good catch!

ashkan-software commented:
HI @vivekmig. Thank you for your comments. I addressed your comments in the last commit. I hope I was able to address them all, as I was not sure about some of them :)

vivekmig commented on captum/attr/_core/neuron/neuron_deep_lift.py:
```diff
@@ -202,13 +209,15 @@ def __init__(self, model, layer):
+        inputs: TensorOrTupleOfTensors,
+        neuron_index: Union[int, Tuple[int, ...]],
+        baselines: Optional[
+            Union[int, float, Tensor, Tuple[Union[Tensor, int, float], ...]]
```
This baselines (only for NeuronDeepLiftShap) Union should also include Callable[..., Union[Tensor, Tuple[Tensor,...]]]

ashkan-software replied:
Thanks!

vivekmig commented on tests/attr/neuron/test_neuron_deeplift_basic.py:
```diff
@@ -114,8 +118,22 @@ def test_relu_deepliftshap_with_custom_attr_func(self):
+            baselines: Optional[
+                Union[Tensor, int, float, Tuple[Union[Tensor, int, float], ...]]
+            ] = None,
+        ) -> TensorOrTupleOfTensors:
```
This return type should be Tuple[Tensor, ...]

ashkan-software replied:
Thanks!

vivekmig commented on tests/attr/neuron/test_neuron_deeplift_basic.py:
```diff
@@ -114,8 +118,22 @@ def test_relu_deepliftshap_with_custom_attr_func(self):
-        def custom_attr_func(multipliers, inputs, baselines):
+    def _relu_custom_attr_func_assert(
+        self,
+        attr_method: Any,
```
This method should have type Union[NeuronDeepLift, NeuronDeepLiftShap]

ashkan-software commented:
@vivekmig Thank you!

It seems like I cannot merge this branch, as I am not authorized....
Also, I do not see any button for importing to Phabricator. Should this button be on this website somewhere?
